src/ir/module.rs

```diff
@@ -327,11 +330,30 @@ impl<'a> Module<'a> {
 
     /// Return a wrapper that implements Display on this module,
     /// pretty-printing it as textual IR.
-    pub fn display<'b>(&'b self) -> ModuleDisplay<'b>
+    pub fn display<'b>(&'b self) -> ModuleDisplay<'b, NOPPrintDecorator>
     where
         'b: 'a,
     {
-        ModuleDisplay { module: self }
+        ModuleDisplay {
+            module: self,
+            decorators: None,
+        }
+    }
+
+    /// Return a wrapper that implements Display on this module,
+    /// pretty-printing it as textual IR with some additional text whose
+    /// printing is described in Decorator.
+    pub fn display_with_decorators<'b, PD: PrintDecorator>(
+        &'b self,
+        decorators: HashMap<Func, &'b PD>,
```

Comment:
I can't have it be a `FnMut` since `Display` only takes in a `&self`. 
So my `ModuleDisplay` now looks like below. I made the function return an `Option` so I could pass in `None` for all functions in the normal display function.  
```
pub struct ModuleDisplay<'a, PD: PrintDecorator, F: Fn(Func) -> Option<PD>> {
    pub(crate) module: &'a Module<'a>,
    pub(crate) decorators: F,
}
```
However, I'm having trouble implementing the display function for Module in a way that doesn't expose these type parameters to the user. Ideally I am able to have both the PrintDecorator and function returning the decorator be impl, like so: 
```
pub fn display<'b>(
    &'b self,
) -> ModuleDisplay<'b, impl PrintDecorator, impl Fn(Func) -> Option<impl PrintDecorator>>
where
    'b: 'a,
{
    ModuleDisplay::<NOPPrintDecorator, dyn Fn(Func) -> Option<NOPPrintDecorator>> {
        module: self,
        decorators: |x: Func| None,
    }
}
``` 
However, this doesn't work since `the trait Sized is not implemented for dyn Fn(ir::Func) -> std::option::Option<display::NOPPrintDecorator>`. I am able to get no type errors by exposing the generics to the user but ideally we don't want that...

Reply:
The `display_with_decorator` looks like this: 
```
pub fn display_with_decorator<'b, PD: PrintDecorator, F: Fn(Func) -> Option<PD>>(
    &'b self,
    decorators: F,
) -> ModuleDisplay<'b, PD, F>
where
   'b: 'a,
{
    ModuleDisplay {
        module: self,
        decorators: decorators,
    }
}
```

Reply:
Does the first variant above work without the type ascription (`ModuleDisplay { ... }` expression only)? Or alternately, `impl Fn` rather than `dyn Fn`?

We're definitely getting into the weeds of the Rust type system here but there has to be a way to make this work!

Reply:
So `impl Fn` doesn't work because `impl` isn't allowed in paths (? idk what a path is). `impl Trait` is only allowed in arguments and return types of functions and methods. 

Removing the type ascription works, however, we now get the error `expected Option<impl PrintDecorator>, found Option<dyn PrintDecorator>`. I think the problem is that it doesn't know that the function we're passing in will return a PrintDecorator which we have already impl'd. 
```
pub fn display<'b>(
    &'b self,
) -> ModuleDisplay<'b, impl PrintDecorator, impl Fn(Func) -> Option<PrintDecorator>>
where
    'b: 'a,
{
    ModuleDisplay {
        module: self,
        decorators: |x: Func| -> Option<NOPPrintDecorator> { None },
    }
}  
```

Reply:
OK, one more idea: can we use a `Box<dyn Fn(...) -> ...>` type for `decorators` and avoid the type parameter in `ModuleDisplay` altogether?

Reply:
(or `Option<Box<dyn Fn...>>` if we want to avoid the allocation in the common case)

Reply:
It works!! Or at least it type checks. I'll push after I test it against my solver.

Reply:
Verified this works with my generators.